clients/src/main/java/org/apache/kafka/common/protocol/DataOutputStreamWritable.java

```diff
@@ -99,7 +99,7 @@ public void writeUnsignedVarint(int i) {
     public void writeByteBuffer(ByteBuffer buf) {
         try {
             if (buf.hasArray()) {
-                out.write(buf.array(), buf.position(), buf.limit());
+                out.write(buf.array(), buf.arrayOffset(), buf.limit());
```

Comment:
Should it be `buf.position() + buf.arrayOffset()`?
> If this buffer is backed by an array then buffer position p corresponds to array index p + arrayOffset(). 

https://docs.oracle.com/javase/8/docs/api/java/nio/Buffer.html#arrayOffset--

Reply:
Yes, it looks like it should be `buf.position() + buf.arrayOffset()`. In addition, `buf.limit()` should be replaced with `buf.remaining()`.

Reply:
Nice catch. Can we extend the tests to cover both of the cases discussed here too?

Reply:
Great catch, thank you for your reviews. I'll fix the implementation and make sure the missed cases are covered by tests.